src/util.ts

```diff
@@ -0,0 +1,191 @@
+  return false;
+}
+
+/**
+ * @public
+ */
+export function selectionCell(
+  state: EditorState,
+): ResolvedPos | null | undefined {
+  const sel = state.selection as CellSelection | NodeSelection;
+  if ('$anchorCell' in sel && sel.$anchorCell) {
+    return sel.$anchorCell.pos > sel.$headCell.pos
+      ? sel.$anchorCell
+      : sel.$headCell;
+  } else if (
+    'node' in sel &&
+    sel.node &&
+    sel.node.type.spec.tableRole == 'cell'
+  ) {
+    return sel.$anchor;
+  }
+  return cellAround(sel.$head) || cellNear(sel.$head);
+}
+
+function cellNear($pos: ResolvedPos): ResolvedPos | undefined {
+  for (
+    let after = $pos.nodeAfter, pos = $pos.pos;
+    after;
+    after = after.firstChild, pos++
+  ) {
+    const role = after.type.spec.tableRole;
+    if (role == 'cell' || role == 'header_cell') return $pos.doc.resolve(pos);
+  }
+  for (
+    let before = $pos.nodeBefore, pos = $pos.pos;
+    before;
+    before = before.lastChild, pos--
+  ) {
+    const role = before.type.spec.tableRole;
+    if (role == 'cell' || role == 'header_cell')
+      return $pos.doc.resolve(pos - before.nodeSize);
+  }
+}
+
+/**
+ * @public
+ */
+export function pointsAtCell($pos: ResolvedPos): boolean {
+  return $pos.parent.type.spec.tableRole == 'row' && !!$pos.nodeAfter;
+}
+
+/**
+ * @public
+ */
+export function moveCellForward($pos: ResolvedPos): ResolvedPos {
+  return $pos.node(0).resolve($pos.pos + $pos.nodeAfter.nodeSize);
+}
+
+/**
+ * @public
+ */
+export function inSameTable($a: ResolvedPos, $b: ResolvedPos): boolean {
+  return $a.depth == $b.depth && $a.pos >= $b.start(-1) && $a.pos <= $b.end(-1);
+}
+
+/**
+ * @public
+ */
+export function findCell($pos: ResolvedPos): Rect {
+  return TableMap.get($pos.node(-1)).findCell($pos.pos - $pos.start(-1));
+}
+
+/**
+ * @public
+ */
+export function colCount($pos: ResolvedPos): number {
+  return TableMap.get($pos.node(-1)).colCount($pos.pos - $pos.start(-1));
+}
+
+/**
+ * @public
+ */
+export function nextCell(
+  $pos: ResolvedPos,
+  axis: string,
+  dir: number,
+): null | ResolvedPos {
+  const start = $pos.start(-1),
+    map = TableMap.get($pos.node(-1));
+  const moved = map.nextCell($pos.pos - start, axis, dir);
+  return moved == null ? null : $pos.node(0).resolve(start + moved);
+}
+
+/**
+ * @public
+ */
+export function setAttr(
+  attrs: Attrs,
+  name: string,
+  value: unknown,
+): MutableAttrs {
+  const result = {};
+  for (const prop in attrs) result[prop] = attrs[prop];
+  result[name] = value;
+  return result;
+}
+
+/**
+ * @public
+ */
+export function removeColSpan(attrs: Attrs, pos: number, n = 1): Attrs {
+  const result = setAttr(attrs, 'colspan', attrs.colspan - n);
+  if (result.colwidth) {
+    result.colwidth = result.colwidth.slice();
+    result.colwidth.splice(pos, n);
+    if (!result.colwidth.some((w) => w > 0)) result.colwidth = null;
+  }
+  return result;
+}
+
+/**
+ * @public
+ */
+export function addColSpan(attrs: Attrs, pos: number, n = 1): Attrs {
```

Comment:
The old `addColSpan` and `removeColSpan` has `<T>`. We should add `<T>` too to avoid breaking changes.

Reply:
I always end up with a typescript error down the line when I try to add generic. I am not entirely convinced whether we need a generic or not here :thinking: 

Would you have a use-case to help convincing me ?

Reply:
To be honest, I don't feel the generic here is much useful. I only reason I posted my previous comment is to avoid breaking change. To make the code clear, I'm ok to remove the generic here. We have such breaking changes during the migration from `@types/prosemirror-*` to official `prosemirror-*` packages too, so I guess it's not a big deal.